public/app/features/explore/slate-plugins/suggestions.tsx

```diff
@@ -6,43 +6,54 @@ import { Editor as CoreEditor } from 'slate';
 import { Plugin as SlatePlugin } from '@grafana/slate-react';
 import { TypeaheadOutput, CompletionItem, CompletionItemGroup } from 'app/types';
 
-import { QueryField, TypeaheadInput } from '../QueryField';
+import { TypeaheadInput } from '../QueryField';
 import TOKEN_MARK from '@grafana/ui/src/slate-plugins/slate-prism/TOKEN_MARK';
 import { TypeaheadWithTheme, Typeahead } from '../Typeahead';
 
 import { makeFragment } from '@grafana/ui';
 
 export const TYPEAHEAD_DEBOUNCE = 100;
 
+// Commands added to the editor by this plugin.
+interface SuggestionsPluginCommands {
+  selectSuggestion: (suggestion: CompletionItem) => CoreEditor;
+  applyTypeahead: (suggestion: CompletionItem) => CoreEditor;
+}
+
 export interface SuggestionsState {
   groupedItems: CompletionItemGroup[];
   typeaheadPrefix: string;
   typeaheadContext: string;
   typeaheadText: string;
 }
 
-let state: SuggestionsState = {
-  groupedItems: [],
-  typeaheadPrefix: '',
-  typeaheadContext: '',
-  typeaheadText: '',
-};
-
 export default function SuggestionsPlugin({
   onTypeahead,
   cleanText,
   onWillApplySuggestion,
-  syntax,
   portalOrigin,
-  component,
 }: {
   onTypeahead: (typeahead: TypeaheadInput) => Promise<TypeaheadOutput>;
   cleanText?: (text: string) => string;
   onWillApplySuggestion?: (suggestion: string, state: SuggestionsState) => string;
-  syntax?: string;
   portalOrigin: string;
-  component: QueryField; // Need to attach typeaheadRef here
 }): SlatePlugin {
+  let typeaheadRef: Typeahead;
+  let state: SuggestionsState = {
+    groupedItems: [],
+    typeaheadPrefix: '',
+    typeaheadContext: '',
+    typeaheadText: '',
+  };
+  const handleTypeaheadDebounced = debounce(handleTypeahead, TYPEAHEAD_DEBOUNCE);
```

Comment:
We need to debounce inside the closure otherwise multiple instances will get the same debounced function which can create all sorts of problems.